### ISSUE TITLE
ERROR: Output stream closed before all output was written to it.

### ISSUE THREAD

**Issue by Betristor** [Pyomo/pyomo#1997]
### Steps to reproduce the issue

```console
$ command1 [options]
$ command2 [options]
...
```

```
# example.py
import pyomo.environ
...
```

### Error Message
ERROR: Output stream closed before all output was written to it. The following
    was left in the output buffer:
Actually I really don't know whether this comes from pyomo or gurobi.

```console
$ # Output message here, including entire stack trace, if available
```

### Information on your system

Pyomo version: 6.0
Python version: 3.7.9
Operating system: Ubuntu 20.04.2 LTS (GNU/Linux 5.8.0-50-generic x86_64)
How Pyomo was installed (PyPI, conda, source): PyPI
Solver (if applicable): Gurobi

### Additional information

+ This situation happened today, and nothing wrong before, sorry I really don't know how to reproduce it
+ I updated all my packages using pipdate yesterday
+ Nothing changed after I reinstall gurobi

![image](https://user-images.githubusercontent.com/55777837/120187307-34d87680-c247-11eb-97da-31edb914ebcf.png)

Guys, I encountered this problem today when I'm solving a milp problem with pyomo. I specificied my solver as gurobi and linked it with excutable path. But once gurobi logging starts to output this error will occur once and once again blinding me from real log I want to check.

I tried to reinstall gurobi on server but nothing changed. To behonest, I'm not even sure of which part should be responsible for this error: pyomo or gurobi. And I searched both websites with keywords like title states, nothing comes in visual. Although my model is still being solving by gurobi, the logging of ERROR is really annoying and noisy.

Does anyone have ever met this? Any solution?

**Comment by rhshadrach** [comment]
Just upgraded to pyomo 6.0 and using Ipopt as a solver, seeing this message as well.

**Comment by rhshadrach** [comment]
I am getting a full log from Pyomo, even though its reporting I'm not. Are you seeing the same @Betristor? If that's the case, you can use `logging.addFilter` to suppress this particular message, the handler is "pyomo.common.tee".

**Comment by jsiirola** [comment]
The error (or at least the fix) is going to be in Pyomo.  Pyomo 6.0 removed the dependency on pyutilib, and reimplemented certain core functionality natively in Pyomo.  One part of that is how we handle "`tee`ing" solver output.  The error you are seeing is because output coming from the solver hasn't been fully flushed before the `TeeStream` handler closes (it comes from `pyomo/common/tee.py:200`).

We spent a lot of time exercising this locally before releasing Pyomo and haven't seen this behavior, so I am guessing a bit here.  The first thing I would like to try is to edit `pyomo/opt/solver/shellcmd.py` with the following patch:
```diff
diff --git a/pyomo/opt/solver/shellcmd.py b/pyomo/opt/solver/shellcmd.py
index 2c0e244..5bda59e 100644
--- a/pyomo/opt/solver/shellcmd.py
+++ b/pyomo/opt/solver/shellcmd.py
@@ -320,6 +320,8 @@ class SystemCallSolver(OptSolver):
                     timeout=timeout,
                     universal_newlines=True,
                 )
+                t.STDOUT.flush()
+                t.STDERR.flush()
 
             rc = results.returncode
             log = ostreams[0].getvalue()
```

You may also consider explicitly uninstalling pyutilib (`pip uninstall pyutilib`)

To aid in further debugging, can you include more of your driver script?  How are you calling `solve()` (i.e., with output redirection, with `tee=True`, etc.)?

**Comment by rhshadrach** [comment]
I tried to apply @jsiirola's patch but am still seeing the the logging error message. The error message appears before the patch is hit. Here is the call we are using:

    solver.solve(m, tee=True, symbolic_solver_labels=True)

where m is our ConcreteModel.

I will see if I can create a reproducible example I can share.

**Comment by jsiirola** [comment]
@rhshadrach, can you also apply the following patch and post the new error:
```diff
diff --git a/pyomo/common/tee.py b/pyomo/common/tee.py
index 3922361..b351999 100644
--- a/pyomo/common/tee.py
+++ b/pyomo/common/tee.py
@@ -197,9 +197,10 @@ class _StreamHandle(object):
                 stream.flush()
             if written != len(ostring):
                 logger.error(
-                    "Output stream closed before all output was "
+                    "Output stream (%s) closed before all output was "
                     "written to it. The following was left in "
-                    "the output buffer:\n\t%r" % (ostring[written:],))
+                    "the output buffer:\n\t%r" % (
+                        stream, ostring[written:],))
 
 
 class TeeStream(object):
```

(I am interested in which output stream -- the `StringIO` buffer or `sys.stdout` -- is being prematurely closed)

**Comment by jsiirola** [comment]
This is making a lot more sense.  The IPython kernel is doing something very similar to what we are doing to manage redirection of stdout/stderr to buffers.  I suspect that there is something where the IPython buffer is filling up and that thread is not getting a chance to empty it.

**Comment by jsiirola** [comment]
OK.  I think I understand what is going on.  IPython's `OutStream.write()` ([source](https://github.com/ipython/ipykernel/blob/fda115626dd434f49a4d8f8e55080d5fd2c2e2b9/ipykernel/iostream.py#L479)) does not adhere to the Python `TextIOBase`  [API](https://docs.python.org/3/library/io.html#io.TextIOBase.write): the API calls for `write()` to return the number of characters written, but IPython's `OutStream.write()` just returns `None`.  

The following patch should get around this bug, but the real solution will be to get IPython to fix their implementation.
```diff
diff --git a/pyomo/common/tee.py b/pyomo/common/tee.py
index 3922361..248c3b9 100644
--- a/pyomo/common/tee.py
+++ b/pyomo/common/tee.py
@@ -195,11 +195,12 @@ class _StreamHandle(object):
                 written = 0
             if written and not self.buffering:
                 stream.flush()
-            if written != len(ostring):
+            if written is not None and written != len(ostring):
                 logger.error(
-                    "Output stream closed before all output was "
+                    "Output stream (%s) closed before all output was "
                     "written to it. The following was left in "
-                    "the output buffer:\n\t%r" % (ostring[written:],))
+                    "the output buffer:\n\t%r" % (
+                        stream, ostring[written:],))
 
 
 class TeeStream(object):
```

**Comment by Betristor** [comment]
Thanks! Problem solved after upgrading to version 6.0.1.

**Comment by ruaridhw** [comment]
> IPython's OutStream.write() (source) does not adhere to the Python TextIOBase API: the API calls for write() to return the number of characters written, but IPython's OutStream.write() just returns None.

@jsiirola for what it's worth, I found this in Celery too and had to fix it with celery/celery#6791. Whilst it was probably pretty reasonable for you to assume `IO.write()` returns `int` since that is its documented return type, I don't think that was the case in Python2 so there seem to be quite a few libraries wrapping `sys.stdout` and not adhering to that API.
